src/parser/parser_utils.rs

```diff
@@ -0,0 +1,303 @@
+// Original file: "ParserMonad.hs"
+// File auto-generated using Corollary.
+
+#[macro_use]
+use corollary_support::*;
+
+use std::mem;
+use std::rc::Rc;
+use std::boxed::FnBox;
+use std::collections::HashSet;
+use std::iter::FromIterator;
+
+use data::r_list::{Reversed, RList, snoc};
+use data::node::{NodeInfo, CNode};
+use data::position::{Position, Pos};
+use data::input_stream::InputStream;
+use data::ident::Ident;
+use data::name::{Name, NameSupply, new_name_supply};
+use parser::tokens::{CToken, CTokEof, movePosLenOfTok};
+use parser::parser::Parser;
+use syntax::ast::*;
+
+#[derive(Debug)]
+pub struct ParseError(pub (Vec<String>, Position));
+
+impl ParseError {
+    pub fn new(pos: Position, msgs: Vec<String>) -> ParseError {
+        ParseError((msgs, pos))
+    }
+}
+
+pub struct PState {
+    curPos: Position,
+    curInput: InputStream,
+    prevToken: Option<CToken>,
+    savedToken: Option<CToken>,
+    nameSupply: NameSupply,
+    tyidents: HashSet<Ident>,
+    scopes: Vec<HashSet<Ident>>,
+}
+
+pub fn execParser<F, T>(input: InputStream, pos: Position, builtins: Vec<Ident>, names: NameSupply,
+                        do_parse: F) -> Result<(T, NameSupply), ParseError>
+    where F: FnOnce(&mut Parser) -> Result<T, ParseError>
+{
+    let initial_state = PState {
+        curPos: pos,
+        curInput: input,
+        prevToken: None,
+        savedToken: None,
+        nameSupply: names,
+        tyidents: HashSet::from_iter(builtins),
+        scopes: vec![],
+    };
+    let (state, res) = Parser::exec(initial_state, do_parse)?;
+    Ok((res, state.nameSupply))
+}
+
+impl Parser {
+    pub fn getNewName(&mut self) -> Name {
+        self.user.nameSupply.next().unwrap()
+    }
+
+    pub fn setPos(&mut self, pos: Position) {
+        self.user.curPos = pos;
+    }
+
+    pub fn getPos(&self) -> Position {  // TODO borrow
+        self.user.curPos.clone()
+    }
+
+    pub fn addTypedef(&mut self, ident: Ident) {
+        self.user.tyidents.insert(ident);
+    }
+
+    pub fn shadowTypedef(&mut self, ident: &Ident) {
+        self.user.tyidents.remove(ident);
+    }
+
+    pub fn isTypeIdent(&self, ident: &Ident) -> bool {
+        self.user.tyidents.contains(ident)
+    }
+
+    pub fn enterScope(&mut self) {
+        self.user.scopes.insert(0, self.user.tyidents.clone());
+    }
+
+    pub fn leaveScope(&mut self) {
+        if self.user.scopes.is_empty() {
+            panic!("leaveScope: already in global scope");
```

Comment:
Yeah, there are many more panics that should be asserts, and also some (in the lexer) that should be `Result`s.

In the end you should be able to throw any source file at it, and it should never panic. Some fuzzing might be fun :)